### PR TITLE
Resolve lifetime issues related to macro dependencies

### PR DESCRIPTION
**Pull request by slavek-kucera**
Still long way to go, but some of the changes look worth committing.

### CODE REVIEW

File: parser_library/src/workspaces/workspace.cpp [hunk 1]
```diff
@@ -184,48 +253,58 @@ workspace_file_info workspace::parse_file(
...
-            filter_and_close_dependencies_(f->files_to_close(), f);
+        for (auto* component : files_to_parse)
+            filter_and_close_dependencies_(component->m_processor_file->files_to_close(), component->m_processor_file);
     }
 
     return ws_file_info;
 }
 
-workspace_file_info workspace::parse_successful(const processor_file_ptr& f)
+workspace_file_info workspace::parse_successful(processor_file_compoments& comp, workspace_parse_lib_provider& libs)
```
Comment (jirimosinger) [inline parser_library/src/workspaces/workspace.cpp:299]:
```suggestion
workspace_file_info workspace::parse_successful(processor_file_compoments& comp, workspace_parse_lib_provider libs)
```

and using `std::move(...)` for `libs` when `parse_successful` is called on lines 199 and 288?

File: parser_library/src/workspaces/workspace.h [hunk 2]
```diff
@@ -183,7 +195,13 @@ class workspace : public diagnosable_impl, public parse_lib_provider
...
         const std::function<std::vector<std::string>(std::string_view)>& instruction_suggestions);
 
-    std::vector<processor_file_ptr> collect_dependants(const utils::resource::resource_location& file_location) const;
+    std::vector<processor_file_compoments*> collect_dependants(const utils::resource::resource_location& file_location);
+
+    processor_file_compoments& add_processor_file_impl(const utils::resource::resource_location& file);
+    processor_file_compoments* find_processor_file_impl(const utils::resource::resource_location& file);
+    const processor_file_compoments* find_processor_file_impl(const utils::resource::resource_location& file) const;
+    friend struct workspace_parse_lib_provider;
+    workspace_file_info parse_successful(processor_file_compoments& comp, workspace_parse_lib_provider&);
```
Comment (jirimosinger) [inline parser_library/src/workspaces/workspace.h:204]:
```suggestion
    workspace_file_info parse_successful(processor_file_compoments& comp, workspace_parse_lib_provider& libs);
```

File: parser_library/src/workspace_manager_impl.h [hunk 3]
```diff
@@ -356,16 +352,13 @@ class workspace_manager::impl final : public diagnosable_impl
...
-            for (auto consumer : parsing_metadata_consumers_)
-            {
-                consumer->consume_parsing_metadata({ metrics, ws_file_info });
-            }
-        }
+        auto proc_file = ws_path_match(document_uri.get_uri()).find_processor_file(document_uri);
+        if (!proc_file)
+            return;
+
+        parsing_metadata data { proc_file->get_metrics(), ws_file_info };
```
[the remark this answers is not in the document]
Comment (slavek-kucera) [reply parser_library/src/workspace_manager_impl.h:359]: it's three bools... but why not 👍

File: parser_library/src/workspaces/workspace.cpp [hunk 4]
```diff
@@ -120,19 +186,22 @@ void workspace::reparse_after_config_refresh()
...
-        if (!found)
+        auto comp = find_processor_file_impl(fname);
+        if (!comp)
             continue;
+
+        auto& found = comp->m_processor_file;
         details.alternative_config = m_configuration.load_alternative_config_if_needed(fname);
-        if (!found->parse(*this, get_asm_options(fname), get_preprocessor_options(fname), &fm_vfm_))
+        workspace_parse_lib_provider ws_lib(*this, fname);
+        if (!found->parse(ws_lib, get_asm_options(fname), get_preprocessor_options(fname), &fm_vfm_))
```
[the remark this answers is not in the document]
Comment (slavek-kucera) [reply parser_library/src/workspaces/workspace.cpp:196]: by construction, it should be always valid.

**Pull request merged**